st2common/st2common/util/crypto.py

```diff
@@ -13,40 +13,291 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+"""
+Module for handling symmetric encryption and decryption of short text values (mostly used for
+encrypted datastore values aka secrets).
+
+NOTE: In the past, this module used and relied on keyczar, but since keyczar doesn't support
+Python 3, we moved to cryptography library.
+
+symmetric_encrypt and symmetric_decrypt functions except values as returned by the AESKey.Encrypt()
+and AESKey.Decrypt() methods in keyczar. Those functions follow the same approach (AES in CBC mode
+with SHA1 HMAC signature) as keyczar methods, but they use and rely on primitives and methods from
+the cryptography library.
+
+This was done to make the keyczar -> cryptography migration fully backward compatible.
+
+Eventually, we should  move to Fernet (https://cryptography.io/en/latest/fernet/) recipe for
+symmetric encryption / decryption, because it offers more robustness and safer defaults (SHA256
+instead of SHA1, etc.).
+"""
+
 from __future__ import absolute_import
+
+import os
+import json
 import binascii
+import base64
+
+from hashlib import sha1
+
+import six
+
+from cryptography.hazmat.primitives.ciphers import Cipher
+from cryptography.hazmat.primitives.ciphers import algorithms
+from cryptography.hazmat.primitives.ciphers import modes
+from cryptography.hazmat.primitives import hashes
+from cryptography.hazmat.primitives import hmac
+from cryptography.hazmat.backends import default_backend
 
 __all__ = [
+    'KEYCZAR_HEADER_SIZE',
+    'KEYCZAR_AES_BLOCK_SIZE',
+    'KEYCZAR_HLEN',
+
     'read_crypto_key',
+
     'symmetric_encrypt',
-    'symmetric_decrypt'
+    'symmetric_decrypt',
+
+    'cryptography_symmetric_encrypt',
+    'cryptography_symmetric_decrypt',
+
+    # NOTE: Keyczar functions are here for testing reasons - they are only used by tests
+    'keyczar_symmetric_encrypt',
+    'keyczar_symmetric_decrypt',
+
+    'AESKey'
 ]
 
+# Keyczar related constants
+KEYCZAR_HEADER_SIZE = 5
+KEYCZAR_AES_BLOCK_SIZE = 16
+KEYCZAR_HLEN = sha1().digest_size
+
 
-def read_crypto_key(key_path, key_type=None):
+class AESKey(object):
+    """
+    Class representing AES key object.
     """
-    Return the crypto key given a path to key file and the key type.
 
-    :param key_path: Absolute path to file containing crypto key.
-    :type key_path: ``str``
+    aes_key_string = None
+    hmac_key_string = None
+    hmac_key_size = None
+    mode = None
+    size = None
 
-    :param key_type: Type of crypto key.
-    :type key_type: :class:`keyczar.keys.KeyType`
+    def __init__(self, aes_key_string, hmac_key_string, hmac_key_size, mode='CBC', size=128):
+        if mode not in 'CBC':
+            raise ValueError('Unsupported mode: %s' % (mode))
 
-    :rtype: ``str``
```

Comment:
I would raise a warning here instead of an error - UNLESS you are maintaining backwards compatibility.

Some users may want to import old (<128 bit) keys.

Reply:
We never supported smaller keys so I think there is no need for that (our tool always generated 128 bit keys).